app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java

```diff
@@ -26,6 +30,81 @@ public Place fetchPlace(String entityID){
+     */
+    public List<Place> fetchPlaces(final LatLng mapBottomLeft, final LatLng mapTopRight) {
+        class Constraint {
+
+            final double latBegin;
+            final double lngBegin;
+            final double latEnd;
+            final double lngEnd;
+
+            public Constraint(final double latBegin, final double lngBegin, final double latEnd,
+                final double lngEnd) {
+                this.latBegin = latBegin;
+                this.lngBegin = lngBegin;
+                this.latEnd = latEnd;
+                this.lngEnd = lngEnd;
+            }
+        }
+
+        final List<Constraint> constraints = new ArrayList<>();
+
+        if (mapTopRight.getLatitude() < mapBottomLeft.getLatitude()) {
+            if (mapTopRight.getLongitude() < mapBottomLeft.getLongitude()) {
+                constraints.add(
+                    new Constraint(mapBottomLeft.getLatitude(), mapBottomLeft.getLongitude(), 90.0,
+                        180.0));
+                constraints.add(new Constraint(mapBottomLeft.getLatitude(), -180.0, 90.0,
+                    mapTopRight.getLongitude()));
+                constraints.add(
+                    new Constraint(-90.0, mapBottomLeft.getLongitude(), mapTopRight.getLatitude(),
+                        180.0));
+                constraints.add(new Constraint(-90.0, -180.0, mapTopRight.getLatitude(),
+                    mapTopRight.getLongitude()));
+            } else {
+                constraints.add(
+                    new Constraint(mapBottomLeft.getLatitude(), mapBottomLeft.getLongitude(), 90.0,
+                        mapTopRight.getLongitude()));
+                constraints.add(
+                    new Constraint(-90.0, mapBottomLeft.getLongitude(), mapTopRight.getLatitude(),
+                        mapTopRight.getLongitude()));
+            }
+        } else {
+            if (mapTopRight.getLongitude() < mapBottomLeft.getLongitude()) {
+                constraints.add(
+                    new Constraint(mapBottomLeft.getLatitude(), mapBottomLeft.getLongitude(),
+                        mapTopRight.getLatitude(), 180.0));
+                constraints.add(
+                    new Constraint(mapBottomLeft.getLatitude(), -180.0, mapTopRight.getLatitude(),
+                        mapTopRight.getLongitude()));
+            } else {
+                constraints.add(
+                    new Constraint(mapBottomLeft.getLatitude(), mapBottomLeft.getLongitude(),
+                        mapTopRight.getLatitude(), mapTopRight.getLongitude()));
+            }
+        }
+
+        final List<Place> cachedPlaces = new ArrayList<>();
+        for (final Constraint constraint : constraints) {
+            cachedPlaces.addAll(placeDao.fetchPlaces(
+                constraint.latBegin,
+                constraint.lngBegin,
+                constraint.latEnd,
+                constraint.lngEnd
+            ));
+        }
```

Comment:
There might be more optimal data structures, such as https://github.com/Quadtree-org/Quadtree or https://github.com/ccperkins/FeSimpleGeoProx or maybe even something provided by Android. This could be the topic of a separate issue. :-)

Reply:
Cool, now that I know about these, my implementation seems quite impetuous in hindsight.

There are atleast two possible approaches to implementing these:

1. _(The easy one):_ Store the places in the room database in the same manner as they currently are. Then, when nearby loads, load each place (or rather just its entityId) into a quadtree (or another data structure) (into memory?). This will still require querying the entire db once, but is much better than the existing implementation.
2. _(The other way, haven't figured out how to do this yet):_ Store the places in the room db in a manner that intrinsically supports fast querying (like storing the quadtree hierarchy).

Which approach do you favor?

Reply:
Approach 2 would be better. QuadTree-org does not have persistance, and neither FeSimpleGeoProx [I believe](https://github.com/ccperkins/FeSimpleGeoProx/issues/2).

I did a few tests and performance does not seem to be an issue so far, but I will create an issue if it does. Thanks! :-)